plugins/inputs/jenkins/jenkins.go

```diff
@@ -0,0 +1,438 @@
+			return badFormatErr(url, tempSpace["size"], "float64", tempSpaceMonitorRoute+".size")
+		}
+	}
+
+	if swapSpaceMonitor := info.MonitorData.Hudson_NodeMonitors_SwapSpaceMonitor; swapSpaceMonitor != nil {
+		swapSpaceMonitorRouter := "hudson.node_monitors.SwapSpaceMonitor"
+		var swapSpace map[string]interface{}
+		if swapSpace, ok = swapSpaceMonitor.(map[string]interface{}); !ok {
+			return badFormatErr(url, swapSpaceMonitor, "map[string]interface{}", swapSpaceMonitorRouter)
+		}
+		if fields["swap_available"], ok = swapSpace["availableSwapSpace"].(float64); !ok {
+			return badFormatErr(url, swapSpace["availableSwapSpace"], "float64", swapSpaceMonitorRouter+".availableSwapSpace")
+		}
+		if fields["swap_total"], ok = swapSpace["totalSwapSpace"].(float64); !ok {
+			return badFormatErr(url, swapSpace["totalSwapSpace"], "float64", swapSpaceMonitorRouter+".totalSwapSpace")
+		}
+		if fields["memory_available"], ok = swapSpace["availablePhysicalMemory"].(float64); !ok {
+			return badFormatErr(url, swapSpace["availablePhysicalMemory"], "float64", swapSpaceMonitorRouter+".availablePhysicalMemory")
+		}
+		if fields["memory_total"], ok = swapSpace["totalPhysicalMemory"].(float64); !ok {
+			return badFormatErr(url, swapSpace["totalPhysicalMemory"], "float64", swapSpaceMonitorRouter+".totalPhysicalMemory")
+		}
+	}
+	acc.AddFields(measurementNode, fields, tags)
+
+	return nil
+}
+
+func (j *Jenkins) gatherNodesData(acc telegraf.Accumulator) {
+	var nodes []*gojenkins.Node
+	var err error
+	err = j.doGet(func() error {
+		nodes, err = j.instance.GetAllNodes()
+		return err
+	})
+
+	url := j.URL + "/computer/api/json"
+	// since gojenkins lib will never return error
+	// returns error for len(nodes) is 0
+	if err != nil || len(nodes) == 0 {
+		acc.AddError(fmt.Errorf("error retrieving nodes[%s]: %v", url, err))
+		return
+	}
+	// get node data
+	for _, node := range nodes {
+		te := j.gatherNodeData(node, url, acc)
+		if te == nil {
+			continue
+		}
+		acc.AddError(te)
+	}
+}
+
+func (j *Jenkins) gatherJobs(acc telegraf.Accumulator) {
+	jobs, err := j.instance.GetAllJobNames()
+	if err != nil {
+		acc.AddError(fmt.Errorf("error retrieving jobs[%s]: %v", j.URL, err))
+		return
+	}
+	var wg sync.WaitGroup
+	for _, job := range jobs {
+		wg.Add(1)
+		go func(name string, wg *sync.WaitGroup, acc telegraf.Accumulator) {
+			defer wg.Done()
+			if te := j.getJobDetail(jobRequest{
+				name:    name,
+				parents: []string{},
+				layer:   0,
+			}, wg, acc); te != nil {
+				acc.AddError(te)
+			}
+		}(job.Name, &wg, acc)
+	}
+	wg.Wait()
+}
+
+// wrap the tcp request with doGet
+// block tcp request if buffered channel is full
+func (j *Jenkins) doGet(tcp func() error) error {
+	j.semaphore <- struct{}{}
+	if err := tcp(); err != nil {
+		<-j.semaphore
+		return err
+	}
+	<-j.semaphore
+	return nil
+}
+
+func (j *Jenkins) getJobDetail(sj jobRequest, wg *sync.WaitGroup, acc telegraf.Accumulator) error {
```

Comment:
Instead of passing waitgroups, would you add a local variable to this function?

```var wg sync.WaitGroup```

and include the wg.Wait

The reason is that this function assumes it is called from the function that does a wait.  You'll want to decouple this.

Reply:
👍  done, not sure I like it though. But since it will match our previous model, the code more consistent for future changes.